Aggressive slicer

polgreen opened the pull request:
This pull request is dependent on:
https://github.com/diffblue/cbmc/pull/1586
https://github.com/diffblue/cbmc/pull/1509

The aggressive slicer by default removes function bodies of any functions not on the shortest path on the call graph between the start function and the property. If no property is specified, it will preseve the shortest path for each property. It is parameterisable:

 --call-depth <n> "    preserves all functions within <n> function calls of the shorets path " --preserve-function <f>      forces the aggressive slicer to preserve function <f>
 --preserve-function containing <f>  force the aggressive slicer to preserve all functions with names containing <f>
 --preserve-all-direct-paths  force the aggressive slicer to preserve all functions on direct paths to the property. This option must be used with a specified property. 

The aggressive slicer is designed to be used in conjunction with:
https://github.com/diffblue/cbmc/pull/1585 - to over-approximate any function bodies we have removed
https://github.com/diffblue/cbmc/pull/1566 - to block functions we do not wish to be included in paths

The results are not sound, it may produce spurious traces, due to over-approximating the removed function bodies, and it may miss traces due to the removed function bodies writing to global variables, or if "preserve-all-direct-paths" is not used. However, it is designed for use with code bases that are far too big to otherwise use cbmc on.

The parameterisation is intended to be used as a means of providing engineer feedback.

polgreen committed:
grapht::output_dot needs digraph declaration

Without the "digraph call_graph{" this is not parsable dot. This change makes the output dot consistent with the previous output_dot function in call_grapht

polgreen committed:
return shortest path from function call graph

Returns the list of functions on the shortest path from a src function to a destination function on the function call graph.

polgreen committed:
Find source location from a property irep_idt

returns the source_locationt for a property irep_idt

polgreen committed:
CR fixes for grapht

Michael's comments on CR-766007 highlighted things that should be changed in grapht.

polgreen committed:
Return functions reachable within N function calls

Adds call_grapht::reachable_within_n_steps. Function is passed an unordered set of function names and a number of steps, N,, and adds any function reachable within N function calls from the original set of functions, to the set of function names

polgreen committed:
Aggressive slicer: automatically stub functions not on shortest path

Automatically removes the function bodies of any function not on the shortest path in the function call graph or reachable within N function calls of the shortest path

polgreen committed:
Introduce reachable call graph class, and rewrite call graph to inherit from grapht<T>

I have rewritten call_grapht to inherit from grapht, because I think there should be fewer graph classes in CBMC.

Instead of constructing the entire call graph, the reachable call graph class constructs only the reachable call graph.

There is some duplication between the call_graph class and the show_call_sequences function, however the show_call_sequences function outputs the graph on the fly, instead of constructing it, which doesn't allow for further use of the graph

polgreen committed:
Aggressive slice should also do a reachability slice and a slice of gobal inits

polgreen committed:
Sound aggressive slicer

Extends the aggressive slicer with an option --preserve-all-direct-paths

This preserves all functions on direct, loop free, paths from the start function to the function containing the property, and removes the function bodies of all other functions.

This significantly reduces the size of the binary whilst ensuring that no counter-examples (possible with the given unwinding limit) are missed (although the counterexamples produced may still be spurious)

polgreen committed:
Documentation of call graph and reachable call graph

polgreen committed:
white space call graph

polgreen commented:
The clang format patch for this pull request would change a lot of code that I haven't touched, e.g., the GOTO_INSTRUMENT_OPTIONS. Should I apply this patch or not? Is there a way to selectively run clang-format on only sections of the file that are changed?

smowton commented:
AFAIK no, but there is such a facility for our own linter -- see `scripts/travis_lint.sh`, which uses `run_diff.sh` to achieve this by running the linter per file and then filtering by git-diff. That script could be adapted to filter clang-format warnings instead of linter warnings? It wouldn't filter the patch produced by git-clang-format, but rather the line-by-line warnings produced by clang-format.

martin-cs commented:
@polgreen clang format should only look at the code in the diff.  If not then something is wrong.  Please don't fix it by just applying the diff it says!

martin-cs commented:
A few general observations:

1. This sounds like an interesting bit of functionality but one that requires a combination of options / passes to use as intended.  Please could we have an addition to the manual that describes it.

2. No chance of a regression test or two to go with it is there?

3. As a general rule, new bits of code are easier to merge if we can be sure that they are independent of existing functionality.  I suspect that to be the case here; can you help with the argument of why?

martin-cs commented on src/analyses/call_graph.cpp:
```diff
@@ -18,14 +18,9 @@ call_grapht::call_grapht()
-}
-
-call_grapht::call_grapht(const goto_functionst &goto_functions)
-{
```
Why remove this constructor?

polgreen replied:
Ah, this is left over from call graph changes. Will revert

martin-cs commented on src/analyses/call_graph.cpp:
```diff
@@ -51,50 +46,158 @@ void call_grapht::add(
+{
+  for(node_indext n = 0; n < nodes.size(); n++)
+    output_xml_node(out, n);
 }
```
Would asking for an output_json be unreasonable?

polgreen replied:
Not unreasonable, it should also be added to grapht, which doesn't have one either.

martin-cs replied:
That'd be lovely.

martin-cs commented on src/analyses/call_graph.cpp:
```diff
@@ -51,50 +46,158 @@ void call_grapht::add(
   }
+
+  return result;
 }
```
How much work would it be to move these algorithms out into util/graph.h.  It seems like this might not be the only place we want to do a shortest path or bounded depth search.

polgreen replied:
There is already a branch by Smowton that combines my call graph pull request and his, so I think I should probably incorporate this change into that.

martin-cs replied:
OK.  If the two of you get together on this I'd be happy to review.

martin-cs commented on src/analyses/reachable_call_graph.cpp:
```diff
@@ -0,0 +1,105 @@
+    }
+  }
+  INVARIANT(!worklist.empty(), "destination function not found");
+
```
This feels a little more like a precondition or possibly a viable user error and thus not an invariant.

martin-cs commented on src/analyses/reachable_call_graph.h:
```diff
@@ -0,0 +1,41 @@
+#include "call_graph.h"
+
+class reachable_call_grapht: public call_grapht
+{
```
Seems a reasonable use of inheritance.

martin-cs commented on src/goto-instrument/goto_instrument_parse_options.cpp:
```diff
@@ -648,17 +650,35 @@ int goto_instrument_parse_optionst::doit()
       else
-        call_graph.output(std::cout);
+        call_graph.output_dot(std::cout);
 
```
In some ways it would be nice to keep output as text, output_dot as dot (and requiring the dot option), output_xml as xml (and requiring the xml option) and so on.

martin-cs commented on src/goto-instrument/goto_instrument_parse_options.h:
```diff
@@ -79,7 +80,12 @@ Author: Daniel Kroening, kroening@kroening.com
-
+  "(aggressive-slice)" \
+  "(call-depth):" \
+  "(harness-generator):" \
```
I may have missed it, but what does this option do?

polgreen replied:
Ah, that shouldn't be in this pull request.

martin-cs replied:
OK.

martin-cs commented on src/goto-instrument/remove_function.h:
```diff
@@ -32,4 +36,70 @@ void remove_functions(
+/// and those that contain a given irep_idt snippet
+/// If no properties are set by the user, we preserve all functions on
+/// the shortest paths to each property.
+class aggressive_slicert
```
Maybe this should have it's own header / implementation file.

martin-cs commented on src/goto-programs/slice_global_inits.cpp:
```diff
@@ -44,26 +41,8 @@ void slice_global_inits(goto_modelt &goto_model)
-  while(!worklist.empty());
+  std::unordered_set<irep_idt, irep_id_hash> functions_reached=
+      call_graph.reachable_functions(entry_point);
 
```
Are there other uses of the callgraph like this?  It seems finding reachable functions might be a common thing to do.

polgreen replied:
As far as I can find, this is the only bit of code that uses this using a graph.

show_call_sequences outputs the function calls, but does it on the fly without constructing the graph. 

find_used_functions obviously needs to compute the reachable functions, but again does it on the fly without constructing a graph.

martin-cs replied:
OK.  I guess I was suggesting that it should become an API and maybe in time these / others could be encouraged to use it.

martin-cs reviewed (changes requested):
Should be OK.

polgreen commented:
@martin-cs 

yes will write something for the manual. Have emailed you about regression tests, not sure how to write tests for goto-instrument. I have some tests ready, but I don't know how to turn them into something that fits in the test framework.

RE independence: do you mean that it doesn't duplicate functionality of existing features? I implemented this because none of the existing features would scale to the code I was looking at; we tried various combinations of the reachability slicer, the full-slicer, the global-init slicer, and nothing produced a binary small enough for cbmc to analyse. 

Or do you mean that it will work, independently of changes to existing functionality of CBMC? I have implemented this so that when called, goto-instrument does a slice of global-inits before performing the aggressive-slice and a reachability-slice afterwards. I could remove that feature, but it wouldn't relaly make sense to use the aggressive slicer without running the reachability slicer afterwards, as it creates a lot of unreachable code.

polgreen committed:
linter fix: line length

polgreen committed:
Aggressive slicer regression tests

polgreen committed:
re-introduce constructor

polgreen commented:
There seem to be two versions of the manual:
http://www.cprover.org/cbmc/doc/manual.pdf - the old one, but this does actually mention slicing. 
http://www.cprover.org/cprover-manual/  - I guess that this is the up to date version, but it doesn't have any documentation of the existing slicing options

I assume I should add to the second one by editing https://github.com/diffblue/cbmc/blob/develop/doc/cprover-manual.md?

martin-cs commented:
http://www.cprover.org/cprover-manual/ is the correct one, which I believe to be generated from
https://github.com/diffblue/cbmc/blob/develop/doc/cprover-manual.md by processes unknown to me.

Regression tests : discussed offline.

Independence : what I mean is "does this patch set alter the functionality of existing code or only add to it".  If it is clearly only new functionality (which, one would hope, is different and distinct to existing things) then it is easier to review as we don't have to worry about breaking things.
